lambdalifting: improve generated environment setup code

zerbina opened the pull request:
## Summary

Use an assignment with an object constructor expression instead of a
call to `internalNew`. This makes the resulting code easier to reason
about for the analysis passes, and the `internalNew` definition in the
`system` module becomes obsolete.

## Details

The csources compiler still needs the `internalNew` symbol, so removing
the definition is only possible after the next csources update.

Because of using an assignment, the environment setup code now looks
like
```nim
:env = Env()
```
instead of
```nim
`=destroy`(:env)
internalNew(:env)
```
after the `injectdestructors` pass has run.

zerbina committed:
lambdalifting: improve generated environment setup code

Summary
=======

Use an assignment with an object constructor expression instead of a
call to `internalNew`. This makes the resulting code easier to reason
about for the analysis passes, and the `internalNew` definition in the
`system` module becomes obsolete.

Details
=======

The csources compiler still needs the `internalNew` symbol, so removing
the definition is only possible after the next csources update.

Because of using an assignment, the environment setup code now looks
like
```nim
:env = Env()
```
instead of
```nim
`=destroy`(:env)
internalNew(:env)
```
after the `injectdestructors` pass has run.

saem commented on compiler/sem/lambdalifting.nim:
```diff
@@ -148,10 +148,10 @@ const
-  result.add newSymNode(a)
-  result.add b
+proc newObjConstr(typ: PType, info: TLineInfo): PNode =
+  ## Creates an object constructor node with no arguments for type `typ`, using
+  ## `info` for source line information.
```
```suggestion
  ## Creates an object construction node with no arguments for type `typ`,
  ## using `info` for source line information.
```

Very soft suggestion to clarify that it's a construction call/expression as opposed to a constructor definition.

zerbina replied:
Yep, that's better, thank you. I think being precise here is important.

saem commented on lib/system.nim:
```diff
@@ -267,6 +267,7 @@ const ThisIsSystem = true
 
 proc internalNew*[T](a: var ref T) {.magic: "New", noSideEffect.}
   ## Leaked implementation detail. Do not use.
+  # TODO: obsolete; remove after the next csource update
```
hmm, did you want to add a check for sym name in magics? I believe that'll allow detection of whether it was an `internalNew` call and that could be an assert (prevent regressions).

zerbina replied:
I went with wrapping the definition in a check for whether `refc` is used. That way, the symbol is only available during the first bootstrap iteration.

saem reviewed (approved):
Nice find, and can confirm that the new assignment based code _is_ easier to read.

zerbina committed:
lambdalifting: clarify doc comment

Co-authored-by: Saem Ghani <saemghani+github@gmail.com>

zerbina committed:
system: remove `internalNew` from compilation

It's only defined for the first bootstrapping iteration, which prevents
the procedure from accidentally being used again.

zerbina commented:
/merge
